rfcs/2020-10-15-3480-file-source-rework.md

```diff
@@ -0,0 +1,425 @@
+# RFC 3480 - 2020-10-15 - File Source Rework
+
+The `file` source is one of Vector's oldest and most widely used components.
+While it began with relatively simple interface and implementation, it has
+collected a large number of knobs, warts, and general issues over time. To
+ensure that this important source continues to serve users well, we should
+invest effort into making it easier to use, extend, and debug.
+
+## Scope
+
+This RFC will cover a general overhaul of the `file` source internals and
+user-facing config. The focus will be on migrating the existing implementation
+to a better structure, not rewriting from scratch.
+
+## Motivation
+
+There are three pages of issues tagged `source: file` in the Vector repo. Below
+I've tried to collect some of the more notable ones and categorize them:
+
+* File source retains files on disk ([#763](https://github.com/timberio/vector/issues/763))
+
+  * Slowness due to many files
+  * Need sequential instead of fair reads
+  * Unclear why files handles were open
+  * Unclear that files were not done being processed
+
+* Checksums are confusing ([#828](https://github.com/timberio/vector/issues/828))
+
+  * Checksums don't work with small files
+
+* `start_at_beginning` is confusing ([#1020](https://github.com/timberio/vector/issues/1020))
+
+  * Config behavior does not match user expectations
+
+* Only warn about small files not empties ([#1065](https://github.com/timberio/vector/issues/1065))
+
+  * Checksums don't work with small files
+  * Poor observability
+
+* Split reads ([#1125](https://github.com/timberio/vector/issues/1125) and [#2992](https://github.com/timberio/vector/issues/2992))
+
+  * Correctness
+
+* Periodically read a non-file file ([#1198](https://github.com/timberio/vector/issues/1198))
+
+  * New mode
+
+* Clean up file checkpoint files ([#1427](https://github.com/timberio/vector/issues/1427))
+
+  * Checkpointing
+
+* File source slows down with many completed files ([#1466](https://github.com/timberio/vector/issues/1466))
+
+  * Slowness due to many files
+  * Read scheduling
+
+* Add path-based fingerprinting ([#1948](https://github.com/timberio/vector/issues/1948))
+
+  * Checksums don't work with small files
+  * inode fingerprinting is fraught
+
+* Log permissions issues ([#2420](https://github.com/timberio/vector/issues/2420))
+
+  * Poor observability
+
+* Stop after reading files (backfill mode) ([#3216](https://github.com/timberio/vector/issues/3216))
+
+  * New mode
+
+* Tailing limited to single core ([#3379](https://github.com/timberio/vector/issues/3379))
+
+  * Performance
+
+* Not releasing descriptors due to falling behind ([#3440](https://github.com/timberio/vector/issues/3440))
+
+  * Need fair reads instead of sequential reads
+  * Unclear why files handles were open
```

Comment:
I'd also add or clarify

* Unclear why file handles for rotated files were open.

It's very common that users include rotated files in their glob patterns. I think we can do better to recognize a rotated file and release its descriptor after a certain amount of inactivity.

Reply:
It's not quite as simple as identifying "this file has been rotated". We could add some tracking of the various paths at which we've observed a file, but I'd rather we just take care of this by not holding open handles to idle files. We can't know that a "rotated" file won't see more writes anyway.